### DISCUSSION TITLE
Building DAMASK on cluster from Source against Centrally Installed PETSc and HDF5

### DISCUSSION THREAD

**Discussion by Carldkennedy** [damask-multiphysics/DAMASK#11] [Q&A] [answered]
We're encountering issues while attempting to build DAMASK against our centrally installed versions of PETSC and HDF5. Our aim is to utilise these centrally installed modules as they are built against our MPI setup. However, during the build we are encountering errors such as `Fatal Error: Cannot open module file ‘hdf5.mod’ for reading at (1): No such file or directory`
Are there any gotchas we should be aware when building DAMASK this way? Any guidance or help would be greatly appreciated.

**Comment by eisenlohr** [comment]
Can you confirm that the Fortran bindings have been included when building your HDF5 library?

**Reply by Carldkennedy** [reply]
Hi Eisenlohr, thank you for the quick response
```
nm -gC ..../HDF5/1.13.3-gompi-2022a/lib64/libhdf5.so | grep -i fortran    
0000000000435e40 D H5T_FORTRAN_S1_g
```

**Reply by eisenlohr** [reply]
Could you try to locate the `hdf5.mod` file in your HDF5 directory structure and possibly add the directory containing it to the "include" locations for your compile step?

**Comment by MarDiehl** [comment, marked as answer]
> Are there any gotchas we should be aware when building DAMASK this way? Any guidance or help would be greatly appreciated.

DAMASK relies on the PETSc build system and PETSc needs to be aware of HDF5 incl. Fortran bindings (see https://damask.mpie.de/support/FAQ/installation#why-does-compilation-of-grid-or-mesh-solver-fail-due-to-missing-files).
If your PETSc is not aware of HDF5 (and FFTW) but you can ensure compatibility, installation requires manual tweaking of `CMakeLists.txt` (the version in the root directory, not `src/CMakeLists.txt`). This file contains some inactive code marked as "experimental" that uses the _FindPkgConfig_ feature of CMake to search for those packages.
Manual modifications are also possible, but for non-standard locations this becomes tedious. The patch for the Debian/Ubuntu package, which relies on HDF5 in a default location, is attached as [add_library_path_no_petsc.txt](https://github.com/eisenforschung/DAMASK/files/15140762/add_library_path_no_petsc.txt)

> ```
> nm -gC ..../HDF5/1.13.3-gompi-2022a/lib64/libhdf5.so | grep -i fortran    
> 0000000000435e40 D H5T_FORTRAN_S1_g
> ```
The Fortran code of HDF5 is in separate `*.so` files called `libhdf5_fortran.so` and `libhdf5_hl_fortran.so` and/or `libhdf5hl_fortran.so`. But the issue here are missing Fortran module files that need to be in the include path, not the missing libraries.
